packages/flutter/lib/src/material/notches.dart

```diff
@@ -0,0 +1,174 @@
+  /// Abstract const constructor. This constructor enables subclasses to provide
+  /// const constructors so that they can be used in const expressions.
+  const Notch();
+
+  /// Returns a path for a notch in the outline of a shape.
+  ///
+  /// The path makes a notch in the host shape that can contain the guest shape.
+  ///
+  /// The `host` is the bounding rectangle for the shape into which the notch will
+  /// be applied. The `guest` is the bounding rectangle of the shape for which we
+  /// are creating a notch in the host.
+  ///
+  /// The `start` and `end` arguments are points on the outline of the host shape
+  /// that will be connected by the returned path.
+  ///
+  /// The returned path may pass anywhere, including inside the guest bounds area,
+  /// and may contain multiple subpaths. The returned path ends at `end` and does
+  /// not end with a [Path.close]. The returned [Path] is built under the
+  /// assumption it will be added to an existing path that is at the `start`
+  /// coordinates using [Path.addPath].
+  Path getPath(Rect host, Rect guest, Offset start, Offset end);
+}
+
+/// Computes smooth circular notches.
+///
+/// The notches computed by this class are paths that smoothly connect a top
+/// horizontal edge with a circular arc.
+///
+/// The notch computed by this class is intended to contain a circle bound by a
+/// square of the guest + [notchMargin] size, and will not be smaller than this
+/// circle.
+///
+/// The guest shape is assumed to be a circle (hence the guest rectangle is
+/// assumed to be a square).
+///
+/// See also: [getPath].
+class CircularNotch implements Notch {
+  /// Constructs a [CircularNotch] that can be used for multiple notch
+  /// computations.
+  const CircularNotch({
+    this.notchMargin = 4.0,
+  });
+
+  /// Minimal margin that will be kept from the guest. Assuming that the guest's
+  /// shape is circular.
+  ///
+  /// The notch will be sized to contain a circle with a radius of notchMargin + guestRadius.
+  /// (guestRadius is half the width or height of the guest bounding box).
+  final double notchMargin;
+
+  /// Returns a path that defines a smooth circular notch in the top edge of the [host] shape.
+  ///
+  /// The guest's shape is assumed to be a circle bounded by the [guest]
+  /// rectangle. The host's top edge is assumed to be a horizontal line.
+  ///
+  /// The returned path makes a notch in the host top edge that can contain the
+  /// guest shape, and keep a [notchMargin] margin between the notch and the
+  /// guest.
+  ///
+  /// The `host` is the bounding rectangle for the shape into which the notch will
+  /// be applied. The `guest` is the bounding rectangle of the shape for which we
+  /// are creating a notch in the host.
+  ///
+  /// The `start` and `end` arguments are points on the outline of the host shape
+  /// that will be connected by the returned path.
+  @override
+  Path getPath(Rect host, Rect guest, Offset start, Offset end) {
```

Comment:
Tests should verify getPath(h, g, start, end) produces the same path as getPath(h, g, end, start)

Reply:
done